### PR TITLE
Feature/content farming calculation

### PR DESCRIPTION
**Pull request by PopCastcle**

### CODE REVIEW

File: apps/queues/src/app/social-reward.sheduler.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  constructor(
+    private contentService: ContentServiceV2,
+    private adsService: AdsService,
+    private queue: Model<Queue>,
+  ) {}
+
+  @Cron(CronExpression.EVERY_DAY_AT_10AM)
+  async distributeSocialReward() {
+    //await this.contentService.expireAllFarmedToken();
+    //check if there is queue command to distribute reward
```
Comment (kittipong-castcle) [inline apps/queues/src/app/social-reward.sheduler.ts:45]:
```suggestion
```

File: apps/queues/src/app/social-reward.sheduler.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    const actionQueue = await this.queue.findOne({
+      'payload.action': 'distribute-social-reward',
+      status: QueueStatus.WAITING,
+    });
+    if (actionQueue) {
```
Comment (kittipong-castcle) [inline apps/queues/src/app/social-reward.sheduler.ts:50]:
```suggestion
    if (actionQueue) return;
```

File: libs/database/src/lib/services/ads.service.ts [hunk 3]
```diff
@@ -328,19 +347,29 @@ export class AdsService {
...
     const filters: FilterQuery<AdsCampaign> = createCastcleFilter(
-      { owner: _id },
+      { owner: { $in: ids } },
       {
         sinceId,
         untilId,
       },
     );
-
+    console.log('filter');
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:372]:
```suggestion
```

File: libs/database/src/lib/services/content.service.v2.ts [hunk 4]
```diff
@@ -60,14 +60,19 @@ import {
...
+  ContentFarmingReponse,
+} from '../models/content-farming.model';
 import { Repository } from '../repositories';
 import { Environment } from '@castcle-api/environments';
 import { CastLogger } from '@castcle-api/logger';
 import { InjectQueue } from '@nestjs/bull';
 import { Queue } from 'bull';
 import { HashtagService } from './hashtag.service';
 
+import cdf from 'castcle-cdf';
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/content.service.v2.ts:74]:
```suggestion
import cdf from 'castcle-cdf';
```

File: libs/database/src/lib/services/content.service.v2.ts [hunk 5]
```diff
@@ -804,6 +809,82 @@ export class ContentServiceV2 {
     );
   };
 
+  getUndistributedContentFarmingCDF = async () => {
+    //const prospectContentFarmings = await this.contentFarmingModel.find({isDistributed: {$ne:true}},{ }, {sort: 'startAt:1'} );
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/content.service.v2.ts:813]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 6]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+                },
+              },
+            ],
+          },
+          session,
+        );
+      }
+    }
+
+    //transfer rewardContent from ads credit lock to cfs[0].account;
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:836]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 7]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+      session,
+    );
+  };
+
+  distributeContentCreatorReward = async (
+    adsplacement: AdsPlacement,
+    reward: AdsSocialReward,
+    session: mongoose.ClientSession,
+  ) => {
+    //!!! TODO should embed in adsplacement
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:674]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 8]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+        $in: adsplacement.contents.map((c) => String(c)),
+      },
+    });
+
+    for (let j = 0; j < adsplacement.contents.length; j++) {
+      const currentCfs = cfs.filter(
+        (c) => String(c.content) === adsplacement.contents[j].contentId,
+      );
+      if (currentCfs.length > 0) {
+        //has content farming
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:746]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 9]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+        const transferTo: any[] = [];
+        const ledgers: any[] = [];
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:748]:
```suggestion
        const transferTo: type?[] = [];
        const ledgers: type?[] = [];
```

File: libs/database/src/lib/services/ads.service.ts [hunk 10]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+    const creatorPool = reward.creatorShare + reward.farmingShare;
+    const creatorRewardPerContent = creatorPool / adsplacement.contents.length;
+    const cfs = await this._contentFarmingModel.find({
+      content: {
+        $in: adsplacement.contents.map((c) => String(c)),
+      },
+    });
+
+    const transferTo: any[] = [];
+    const ledgers: any[] = [];
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:684]:
```suggestion
    const transferTo: type?[] = [];
    const ledgers: type?[] = [];
```

File: libs/database/src/lib/services/ads.service.ts [hunk 11]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+  };
+
+  distributeAdsReward = async (
+    adsplacement: AdsPlacement,
+    reward: AdsSocialReward,
+    session: mongoose.ClientSession,
+  ) => {
+    //castRate = 1/castActualCost;
+
+    //distribute Content Farming
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:632]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 12]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+  };
+
+  distributeSocialRewardPersonal = async () => {
+    const adsIncomeToken = await this.taccountService.getBalance(
+      CACCOUNT_NO.SOCIAL_REWARD.PERSONAL.NO,
+    );
+    const adsPlacements = await this._adsPlacementModel.find({
+      'cost.CAST': { $exits: false },
+    });
+    //gety cost UST in aggregation
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:602]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 13]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
   };
 
   auctionAds = async (accountId: string) => (await this.getAds(accountId))[0];
+
+  distributeSocialRewardAdsCredit = async () => {
+    //assume all conterm farming has been weigh
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:565]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 14]
```diff
@@ -525,4 +560,279 @@ export class AdsService {
...
+    const adsIncomeToken = await this.taccountService.getBalance(
+      CACCOUNT_NO.SOCIAL_REWARD.ADS_CREDIT.NO,
+    );
+    const adsPlacements = await this._adsPlacementModel.find({
+      'cost.CAST': { $exits: false },
+    });
+    //gety cost UST in aggregation
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:572]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 15]
```diff
@@ -242,12 +253,20 @@ export class AdsService {
     //invalid balance
     if (!(balance / mockOracleService.getCastPrice() >= adsRequest.dailyBudget))
       return false;
+    //have permission to add payment
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:256]:
```suggestion
```

File: libs/database/src/lib/services/ads.service.ts [hunk 16]
```diff
@@ -242,12 +253,20 @@ export class AdsService {
...
+    if (adsRequest.userId && user.id !== adsRequest.userId) {
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/ads.service.ts:257]:
```suggestion
    if (user.id !== adsRequest.userId) {
```

File: libs/database/src/lib/schemas/ads-placement.schema.ts [hunk 17]
```diff
@@ -23,19 +23,23 @@
...
+  AdsCost,
+  AdsPlacementCampaign,
+  AdsPlacementContent,
+} from '../models/ads.model';
 
 @Schema({ timestamps: true })
 export class AdsPlacement extends CastcleBase {
   @Prop({ required: true, type: Array })
-  contents: any[]; // array ob objectId
+  contents: AdsPlacementContent[]; // array ob objectId
```
Comment (kittipong-castcle) [inline libs/database/src/lib/schemas/ads-placement.schema.ts:36]:
```suggestion
  @Prop({ required: true, type: [SchemaTypes.ObjectId] })
  contents: AdsPlacementContent[];
```

File: libs/database/src/lib/schemas/ads-campaign.schema.ts [hunk 18]
```diff
@@ -23,7 +23,7 @@
 
 import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
 import { SchemaTypes } from 'mongoose';
-import { Account, CastcleBase } from '.';
+import { User, CastcleBase } from '.';
```
Comment (kittipong-castcle) [inline libs/database/src/lib/schemas/ads-campaign.schema.ts:26]:
```suggestion
import { CastcleBase } from './common.ts';
import { User } from './user.ts';
```

circular
user -> index -> ads

**Review by kittipong-castcle — CHANGES_REQUESTED** [review]

File: libs/database/src/lib/schemas/ads-placement.schema.ts (hunk 17 above)
Comment (PopCastcle) [reply libs/database/src/lib/schemas/ads-placement.schema.ts]: type is not object id

**Review by nucastcle — APPROVED** [review]

File: apps/queues/src/app/social-reward.sheduler.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    const actionQueue = await this.queue.findOne({
+      'payload.action': 'distribute-social-reward',
+      status: QueueStatus.WAITING,
+    });
+    if (!actionQueue) return;
+    await this.contentService.updateAllUndistributedContentFarming();
+    await this.adsService.distributeSocialRewardAdsCredit();
+    await this.adsService.distributeSocialRewardPersonal();
+    actionQueue.status = QueueStatus.DONE;
+    actionQueue.save();
```
Comment (kittipong-castcle) [inline apps/queues/src/app/social-reward.sheduler.ts:53]:
```suggestion
    await actionQueue.save();
```

File: libs/database/src/lib/services/content.service.v2.ts [hunk 20]
```diff
@@ -804,6 +808,81 @@ export class ContentServiceV2 {
...
+      'farming.isDistributed': { $ne: true },
+    });
+    const result: ContentFarmingCDF[] = [];
+    contents.forEach((item) => {
+      result.push({
+        contentId: item.id,
+        contentFarmings: item.farming,
+      });
+    });
+    return result;
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/content.service.v2.ts:823]:
```suggestion
    return contents.map<ContentFarmingCDF>((content) => ({
      contentId: content.id,
      contentFarmings: content.farming,
    }));
```

**Review by kittipong-castcle — CHANGES_REQUESTED** [review]

**Review by kittipong-castcle — APPROVED** [review]

**Pull request merged**